rasa/core/trackers.py

```diff
@@ -420,11 +432,67 @@ def undo_till_previous(event_type, done_events):
                 # the `action_listen` action right before it).
                 undo_till_previous(UserUttered, applied_events)
                 undo_till_previous(ActionExecuted, applied_events)
+            elif (
+                isinstance(event, ActionExecuted)
+                and event.action_name in form_names
+                and not self._upcoming_rejection(idx)
+                and not self._first_loop_execution_or_previous_rejection(
+                    event.action_name, applied_events
+                )
+            ):
+                self.undo_till_previous_loop_execution(
+                    event.action_name, applied_events
+                )
+            elif isinstance(event, ActionExecutionRejected):
+                # Undo previously executed action
+                undo_till_previous(ActionExecuted, applied_events)
+                applied_events.append(event)
             else:
                 applied_events.append(event)
 
         return applied_events
 
+    def _upcoming_rejection(self, index: int) -> bool:
+        return next(
+            (
+                isinstance(e, ActionExecutionRejected)
```

Comment:
I think it is a good idea to check that ActionExecutionRejected corresponds to the active loop name at that time, should be the name of ActionExecuted for which you look for upcoming rejection